File: kit/dotnet/NerdKey.Kit/Ed25519Verifier.cs
using System.Text;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Signers;

namespace NerdKey.Kit;

/// <summary>
/// Offline Ed25519 signature verification for Keygen ED25519_SIGN license keys.
///
/// Key format: key/&lt;base64url-payload&gt;.&lt;base64url-sig&gt;
/// Signed message: UTF-8 bytes of "key/" + base64url-payload
/// Public key encoding (from Keygen CLI): base64( hex-string( 32-byte raw Ed25519 public key ) )
/// </summary>
public sealed class Ed25519Verifier
{
    private readonly Ed25519PublicKeyParameters _publicKey;

    /// <summary>Initialize using the constant embedded in NerdKeyConstants.</summary>
    public Ed25519Verifier() : this(NerdKeyConstants.Ed25519PublicKeyBase64) { }

    /// <summary>Initialize from a base64-encoded hex-string public key.</summary>
    public Ed25519Verifier(string base64EncodedHexKey)
    {
        var hexBytes = Convert.FromBase64String(base64EncodedHexKey);
        var hexString = Encoding.ASCII.GetString(hexBytes);
        if (hexString.Length != 64)
            throw new InvalidSignatureException();
        var rawKey = Convert.FromHexString(hexString);
        _publicKey = new Ed25519PublicKeyParameters(rawKey, 0);
    }

    /// <summary>
    /// Verify a Keygen ED25519_SIGN license key offline.
    /// Returns the decoded payload JSON on success.
    /// Throws InvalidSignatureException or InvalidLicenseException on failure.
    /// </summary>
    public LicenseKeyPayload Verify(string licenseKey)
    {
        if (!licenseKey.StartsWith("key/", StringComparison.Ordinal))
            throw new InvalidLicenseException("key must start with 'key/'");

        var stripped = licenseKey[4..]; // drop "key/"
        var lastDot = stripped.LastIndexOf('.');
        if (lastDot < 0)
            throw new InvalidLicenseException("key missing signature separator '.'");

        var payloadB64 = stripped[..lastDot];
        var sigB64Url = stripped[(lastDot + 1)..];

        // Message is "key/" + payloadB64 as UTF-8
        var message = Encoding.ASCII.GetBytes("key/" + payloadB64);

        // Decode base64url signature
        var sigBytes = Base64UrlDecode(sigB64Url);

        var verifier = new Ed25519Signer();
        verifier.Init(false, _publicKey);
        verifier.BlockUpdate(message, 0, message.Length);
        if (!verifier.VerifySignature(sigBytes))
            throw new InvalidSignatureException();

        // Decode the payload JSON
        var payloadBytes = Convert.FromBase64String(PadBase64(payloadB64));
        var json = Encoding.UTF8.GetString(payloadBytes);
        var payload = System.Text.Json.JsonSerializer.Deserialize<LicenseKeyPayload>(json);
        if (payload is null)
            throw new InvalidLicenseException("could not parse license key payload");
        return payload;
    }

    private static byte[] Base64UrlDecode(string base64Url)
    {
        return Convert.FromBase64String(PadBase64(
            base64Url.Replace('-', '+').Replace('_', '/')));
    }

    private static string PadBase64(string s)
    {
        int rem = s.Length % 4;
        if (rem == 2) return s + "==";
        if (rem == 3) return s + "=";
        return s;
    }
}

/// <summary>Decoded payload from a Keygen ED25519_SIGN license key.</summary>
public sealed class LicenseKeyPayload
{
    [System.Text.Json.Serialization.JsonPropertyName("license")]
    public LicenseInfo? License { get; set; }

    [System.Text.Json.Serialization.JsonPropertyName("policy")]
    public PolicyInfo? Policy { get; set; }

    public sealed class LicenseInfo
    {
        [System.Text.Json.Serialization.JsonPropertyName("id")]
        public string Id { get; set; } = "";

        [System.Text.Json.Serialization.JsonPropertyName("expiry")]
        public string? Expiry { get; set; }
    }

    public sealed class PolicyInfo
    {
        [System.Text.Json.Serialization.JsonPropertyName("id")]
        public string Id { get; set; } = "";

        [System.Text.Json.Serialization.JsonPropertyName("duration")]
        public int? Duration { get; set; }
    }
}

File: kit/swift/Sources/NerdKeyKit/Ed25519Verifier.swift
import Foundation
import Crypto

/// Offline Ed25519 verification for Keygen ED25519_SIGN license keys.
///
/// Key format:  key/<base64url-payload>.<base64url-sig>
/// The signature is produced over the ASCII bytes of <base64url-payload>.
/// Public key encoding from Keygen CLI: base64( hex-string( 32-byte raw Ed25519 public key ) )
public struct Ed25519Verifier {

    public enum VerificationError: Error {
        case malformedKeyFormat
        case malformedPublicKey
        case signatureInvalid
    }

    private let publicKey: Curve25519.Signing.PublicKey

    /// Load verifier using the constant embedded in Constants.swift.
    public init() throws {
        try self.init(base64EncodedHexKey: NerdKeyConstants.ed25519PublicKeyBase64)
    }

    /// Load verifier from an arbitrary base64-encoded hex-string public key.
    public init(base64EncodedHexKey: String) throws {
        guard
            let hexData = Data(base64Encoded: base64EncodedHexKey),
            let hexString = String(data: hexData, encoding: .utf8)
        else {
            throw VerificationError.malformedPublicKey
        }
        guard hexString.count == 64,
              let rawBytes = Data(hexString: hexString)
        else {
            throw VerificationError.malformedPublicKey
        }
        do {
            publicKey = try Curve25519.Signing.PublicKey(rawRepresentation: rawBytes)
        } catch {
            throw VerificationError.malformedPublicKey
        }
    }

    /// Verify a Keygen license key string offline.
    /// - Returns: Decoded payload JSON data on success.
    /// - Throws: VerificationError on any structural or cryptographic failure.
    @discardableResult
    public func verify(licenseKey: String) throws -> LicenseKeyPayload {
        // Strip "key/" prefix
        guard licenseKey.hasPrefix("key/") else {
            throw VerificationError.malformedKeyFormat
        }
        let stripped = String(licenseKey.dropFirst(4))

        // Split at last "."
        guard let dotRange = stripped.range(of: ".", options: .backwards) else {
            throw VerificationError.malformedKeyFormat
        }
        let payloadB64 = String(stripped[stripped.startIndex..<dotRange.lowerBound])
        let sigB64Url  = String(stripped[dotRange.upperBound...])

        // Keygen ED25519_SIGN: the message is "key/<payload>" (i.e. the full key minus the last ".sig" part)
        // Confirmed by testing: signed over "key/" + payloadB64 as ASCII bytes
        let messageBytes = Data(("key/" + payloadB64).utf8)

        // Decode signature (base64url, padded as needed)
        guard let sigBytes = Data(base64URLEncoded: sigB64Url) else {
            throw VerificationError.malformedKeyFormat
        }

        // Verify
        guard publicKey.isValidSignature(sigBytes, for: messageBytes) else {
            throw VerificationError.signatureInvalid
        }

        // Decode payload JSON
        guard let payloadData = Data(base64Encoded: payloadB64) else {
            throw VerificationError.malformedKeyFormat
        }
        let decoder = JSONDecoder()
        guard let payload = try? decoder.decode(LicenseKeyPayload.self, from: payloadData) else {
            throw VerificationError.malformedKeyFormat
        }
        return payload
    }
}

/// The JSON payload embedded inside a Keygen ED25519_SIGN license key.
public struct LicenseKeyPayload: Decodable {
    public struct LicenseInfo: Decodable {
        public let id: String
        public let created: String
        public let expiry: String?
    }
    public struct PolicyInfo: Decodable {
        public let id: String
        public let duration: Int?
    }
    public let license: LicenseInfo
    public let policy: PolicyInfo
}

// MARK: - Helpers

extension Data {
    /// Decode base64url (with or without padding).
    init?(base64URLEncoded string: String) {
        var base64 = string
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let remainder = base64.count % 4
        if remainder != 0 {
            base64 += String(repeating: "=", count: 4 - remainder)
        }
        self.init(base64Encoded: base64)
    }

    /// Decode a lowercase hex string into bytes.
    init?(hexString: String) {
        let hex = hexString.lowercased()
        guard hex.count % 2 == 0 else { return nil }
        var bytes = [UInt8]()
        bytes.reserveCapacity(hex.count / 2)
        var index = hex.startIndex
        while index < hex.endIndex {
            let nextIndex = hex.index(index, offsetBy: 2)
            guard let byte = UInt8(hex[index..<nextIndex], radix: 16) else { return nil }
            bytes.append(byte)
            index = nextIndex
        }
        self.init(bytes)
    }
}
